include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -464,44 +483,66 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
 template <typename _ContainerOrIterable>
 using __value_t = typename __internal::__memobj_traits<_ContainerOrIterable>::value_type;
 
-template <typename _Accessor>
-struct __usm_or_buffer_accessor
+//-----------------------------------------------------------------------
+// types to create and use data on a device and return those to the host
+//-----------------------------------------------------------------------
+
+template <typename _T, sycl::access_mode _AccessMode>
+struct __combi_accessor
 {
   private:
-    using _T = std::decay_t<typename _Accessor::value_type>;
-    _Accessor __acc;
+    using __acc_t = sycl::accessor<_T, 1, _AccessMode, __dpl_sycl::__target_device, sycl::access::placeholder::false_t>;
     _T* __ptr = nullptr;
-    bool __usm = false;
-    size_t __offset = 0;
+    __acc_t __acc;
 
-  public:
-    // Buffer accessor
-    __usm_or_buffer_accessor(sycl::handler& __cgh, sycl::buffer<_T, 1>* __sycl_buf,
-                             const sycl::property_list& __prop_list)
-        : __acc(*__sycl_buf, __cgh, __prop_list)
-    {
-    }
-    __usm_or_buffer_accessor(sycl::handler& __cgh, sycl::buffer<_T, 1>* __sycl_buf, size_t __acc_offset,
-                             const sycl::property_list& __prop_list)
-        : __acc(*__sycl_buf, __cgh, __prop_list), __offset(__acc_offset)
+    template <bool __with_offset = false>
+    __acc_t
+    __init_acc(bool __fake, sycl::buffer<_T, 1>& __sycl_buf, sycl::handler& __cgh,
+               const sycl::property_list& __prop_list, std::size_t __sz = 0, std::size_t __offset = 0)
     {
+        if (__fake)
+        {
+            return __acc_t(
+#if _ONEDPL_SYCL2020_DEFAULT_ACCESSOR_CONSTRUCTOR_BROKEN
+                __sycl_buf, __cgh, __prop_list
+#endif
+            );
+        }
+        if constexpr (__with_offset)
+            return __acc_t(__sycl_buf, __cgh, sycl::range{__sz}, sycl::id{__offset}, __prop_list);
+        else
+            return __acc_t(__sycl_buf, __cgh, __prop_list);
     }
 
-    // USM pointer
-    __usm_or_buffer_accessor(sycl::handler&, _T* __usm_buf, const sycl::property_list&)
-        : __ptr(__usm_buf), __usm(true)
-    {
-    }
-    __usm_or_buffer_accessor(sycl::handler&, _T* __usm_buf, size_t __ptr_offset, const sycl::property_list&)
-        : __ptr(__usm_buf), __usm(true), __offset(__ptr_offset)
+  public:
+    __combi_accessor(sycl::handler& __cgh, sycl::buffer<_T, 1>& __sycl_buf, _T* __usm_buf,
+                     const sycl::property_list& __prop_list)
+        : __ptr(__usm_buf), __acc(__init_acc((bool)__ptr, __sycl_buf, __cgh, __prop_list))
```

Comment:
1) `(bool)__ptr` will work incorrectly if you change the order of fiends in this structure
2) I believe more readable to write `__ptr != nullptr` or taking into account (1) - `__usm_buf != nullptr`